### PR TITLE
WIP: SAP Hana dialect

### PR DESCRIPTION
**Pull request by ariel-bentu**
Supporting SAP HANA Dialect.

At this point only tables.

TODO:
- [x] Support functions and Views
- [x] Docs

Related to [#212](https://github.com/mtxr/vscode-sqltools/issues/212)
Related to [#20](https://github.com/mtxr/vscode-sqltools/issues/20)

### CODE REVIEW

File: packages/core/dialect/saphana/index.ts [hunk 1]
```diff
@@ -0,0 +1,199 @@
+
+import {
+  ConnectionDialect
+} from '@sqltools/core/interface';
+import * as Utils from '@sqltools/core/utils';
+import queries from './queries';
+import GenericDialect from '@sqltools/core/dialect/generic';
+import { DatabaseInterface } from '@sqltools/core/plugin-api';
+// Cannot simply use hanaClient from '@sap/hana-client', because webpack is moving it away from its native libs.
```
Comment (mtxr) [inline packages/core/dialect/saphana/index.ts]: Take a look at SQLite and Oracle implementation. I believe that this case would be better if handled by the dependencyManager like SQLite and Oracle does

**Comment by mtxr** [comment]
@ariel-bentu nice work! That was fast!

File: packages/core/dialect/saphana/index.ts (hunk 1 above)
Comment (ariel-bentu) [reply packages/core/dialect/saphana/index.ts]: done

File: packages/core/dialect/saphana/index.ts [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+  private resolveQueryResults(resolve, rows, query) {
+    let cols: string[] = [];
+    if (rows && rows.length > 0) {
+      for (let colName in rows[0]) {
+        cols.push(colName);
+      }
+    }
+
+    let res = {
+      connId: "1",
```
Comment (mtxr) [inline packages/core/dialect/saphana/index.ts]:
```suggestion
      connId: this.getId(),
```

This is the conn id generated by extension. It's used for some commands like exporting results

File: packages/core/dialect/saphana/index.ts [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+  }
+
+  private resolveErr(resolve, err, query) {
+    let messages: string[] = [];
+    if (err.message) {
+      messages.push(err.message);
+    }
+
+    return resolve([{
+      connId: "1",
```
Comment (mtxr) [inline packages/core/dialect/saphana/index.ts]:
```suggestion
      connId: this.getId(),
```

File: packages/plugins/package.json [hunk 4]
```diff
@@ -14,7 +14,8 @@
     "@sqltools/language-server": "*",
     "@sqltools/ui": "*",
     "csv-stringify": "^5.3.0",
-    "lodash": "^4.17.11"
+    "lodash": "^4.17.11",
+    "@sap/hana-client": "^2.4.126"
```
Comment (mtxr) [inline packages/plugins/package.json]: We dont need to add it as a dep on package.json anymore.

File: packages/core/package.json [hunk 5]
```diff
@@ -27,6 +27,7 @@
     "@types/redux": "^3.6.0",
     "@types/sqlite3": "^3.1.5",
     "@types/webpack-env": "^1.13.8",
+    "@sap/hana-client": "^2.4.126",
```
Comment (mtxr) [inline packages/core/package.json]: Not needed

File: packages/language-server/webpack.config.js [hunk 6]
```diff
@@ -4,6 +4,9 @@ const outdir = path.resolve(__dirname, '..', '..', '..', 'dist');
 
 const babelOptions = require(path.join(__dirname, '.babelrc'));
 
+const CopyPlugin = require('copy-webpack-plugin');
```
Comment (mtxr) [inline packages/language-server/webpack.config.js]: Not needed

File: packages/language-server/webpack.config.js [hunk 7]
```diff
@@ -26,9 +29,17 @@ module.exports = function getLanguageServerConfig() {
           use: [{ loader: 'babel-loader', options: babelOptions }],
           exclude: /\.test\..+/i,
 
-        },
+        }
       ],
     },
+    plugins: [
```
Comment (mtxr) [inline packages/language-server/webpack.config.js]: Not needed

**Comment by ariel-bentu** [comment]
> On first usage, user will be requested to install it, that helps on packing because it allows us to pack for multi platforms and let node handle deps.

is there a way to advice the dependency manager that this package is?
The package is in a public npm repo maintained by SAP at https://npm.sap.com

Instructions from its site is: 
npm config set @sap:registry https://npm.sap.com
npm install @sap/hana-client

In addition, why is it not finding the module, if I previouslly installed it globally (manually ran `npm install -g ...` )?
Thanks

File: .test-database/.vscode/settings.json [hunk 8]
```diff
@@ -120,5 +120,16 @@
         "authProtocol": "xprotocol"
       }
     },
+    {
```
[the remark this answers is not in the document]
Comment (ariel-bentu) [reply .test-database/.vscode/settings.json]: it is ok, good as an example and not sensitive

File: packages/core/dialect/saphana/index.ts (hunk 3 above)
Comment (ariel-bentu) [reply packages/core/dialect/saphana/index.ts]: adopted

File: packages/core/dialect/saphana/index.ts [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+  exec(query: string, handler: (err: any, row: any) => void);
+  disconnect();
+  prepare(query: string, handler: (err: any, statement: Statement) => void);
+}
+
+export default class SAPHana extends GenericDialect<HanaConnection> implements ConnectionDialect {
+
+  public static deps: typeof GenericDialect['deps'] = [{
+    type: 'package',
+    name: '@sap/hana-client'
```
[the remark this answers is not in the document]
Comment (ariel-bentu) [reply packages/core/dialect/saphana/index.ts]: done

**Comment by mtxr** [comment]
> > On first usage, user will be requested to install it, that helps on packing because it allows us to pack for multi platforms and let node handle deps.
> 
> is there a way to advice the dependency manager that this package is?
> The package is in a public npm repo maintained by SAP at https://npm.sap.com
> 
> Instructions from its site is:
> npm config set @SAP:registry https://npm.sap.com
> npm install @sap/hana-client
> 
> In addition, why is it not finding the module, if I previouslly installed it globally (manually ran `npm install -g ...` )?
> Thanks

No, we don't have it for now. But I'll add in you PR, ok? @ariel-bentu

File: packages/core/dialect/saphana/debug.js [hunk 10]
```diff
@@ -0,0 +1,5 @@
+module.exports = function (ns){
```
Comment (mtxr) [inline packages/core/dialect/saphana/debug.js]: Is this file mandatory? If not, please, consider removing

**Comment by ariel-bentu** [comment]
go ahead and merge. 
any idea why global on npm is not recognized?

File: packages/core/dialect/saphana/index.ts [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+
+  public async close() {
+    if (!this.connection) return Promise.resolve();
+
+    await this.connection.then(conn => conn.disconnect());
+    this.connection = null;
+  }
+
+  public async testConnection?() {
+    return this.open().then(conn => conn.exec('select 1 from dummy;', (err, rows) => rows));
```
Comment (mtxr) [inline packages/core/dialect/saphana/index.ts]: is it a real query? I dont know if dummy exists for all SAPHana dbs, I would suggest to use something like `select 1` or `select CURRENT_DATE`

**Comment by mtxr** [comment]
About the modules, for now it's better to install the module inside of the extension folder because in global install users usually have the latest version, and for the extension we are going to fix it in a tested version.


I'll send a review. It's 2 questions only. See below:

File: packages/core/dialect/saphana/debug.js (hunk 10 above)
Comment (mtxr) [reply packages/core/dialect/saphana/debug.js]: Also, please remove that DS_store file and add it to .gitignore on project root

**Comment by mtxr** [comment]
I'll add the registry specification and reach you back.

File: packages/core/dialect/saphana/index.ts (hunk 11 above)
Comment (ariel-bentu) [reply packages/core/dialect/saphana/index.ts]: yes it is a real query and always there

File: packages/core/dialect/saphana/debug.js (hunk 10 above)
Comment (ariel-bentu) [reply packages/core/dialect/saphana/debug.js]: no actually this file is left over before move to package manager

File: packages/core/dialect/saphana/index.ts (hunk 11 above)
Comment (mtxr) [reply packages/core/dialect/saphana/index.ts]: thanks!

**Comment by mtxr** [comment]
@ariel-bentu  https://github.com/ariel-bentu/vscode-sqltools/pull/1/files please see and test it, but I believe after this we can merge it on master

**Comment by mtxr** [comment]
Is it fully working? @ariel-bentu

**Comment by ariel-bentu** [comment]
> Is it fully working? @ariel-bentu

I sent you by mail, you need to change to:
    args: ['--@sap:registry=https://npm.sap.com']

**Comment by ariel-bentu** [comment]
Regarding the -g option: using a globally installed package instead of requiring installation:

I found that npm has the command link:
```
npm link '@sap/hana-client'
```

It would be nice to first try to link, and if fails (exit code != 0) fall back to installing it. what do you think?

**Pull request merged**